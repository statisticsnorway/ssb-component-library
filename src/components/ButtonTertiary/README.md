# Button Tertiary

> Action button

    - with Accordion functionality
    - with normal onClick functionality

### Usage

#### HTML

```html
<div class="ssb-btn-tertiary">
  <button class="button-header closed icon">
    <span class="button-grid">
      <span class="header-text"> Tertiary button with Icon & Accordion </span>
      <i>{feather.chevronDown 18px}</i>
    </span>
  </button>
  <div class="accordion-body closed"></div>
</div>

<button class="ssb-btn-tertiary">
  <button class="button-header closed no-icon">
    <span class="button-grid">
      <span class="header-text"> Tertiary button with onClick </span>
    </span>
  </button>
</button>

<button class="ssb-btn-tertiary negative">
  <button class="button-header closed no-icon" disabled>
    <span class="button-grid">
      <span class="header-text"> Tertiary button with onClick </span>
    </span>
  </button>
</button>
```

#### React

```jsx harmony

<ButtonTertiary header="Tertiary button with Icon & Accordion">
    <p>Explain something about something with something clever.</p>
</ButtonTertiary>

<ButtonTertiary header="Tertiary button with onClick" accordion={false} onClick={() => testOnClick()} />

<ButtonTertiary header="Knapp negative med Icon & Accordion" negative disabled>
	<p>Explain something about something with something clever.</p>
</ButtonTertiary>

```

Available props:

| Name          | Type     | Description                                                            |
| ------------- | -------- | ---------------------------------------------------------------------- |
| id            | string   | Button text or/and icon                                                |
| children      | node     | Button text or/and icon                                                |
| className     | string   | Optional container class                                               |
| header        | string   | Header text                                                            |
| openByDefault | bool     | Will set the open state on init                                        |
| tabIndex      | number   | Tab index for focus                                                    |
| accordion     | bool     | Will signify button & icon, with an accordion feature on initial state |
| disabled      | bool     | Decides if the button is disabled                                      |
| negative      | bool     | Changes design                                                         |
| onClick       | function | Callback for button click                                              |
