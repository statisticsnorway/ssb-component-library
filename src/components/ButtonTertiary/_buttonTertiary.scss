.ssb-btn-tertiary {
  .button-header {
    @include reset-button;
    @include roboto;
    font-family: 'Open Sans', sans-serif;
    font-style: normal;
    font-weight: 700;
    font-size: 16px;
    line-height: 20px;

    .button-grid {
      cursor: pointer;
      min-height: 44px;
      display: flex;
      align-items: center;
      justify-content: flex-start;
      grid-column-gap: 6px;
    }

    svg {
      color: $ssb-green-4;
      font-size: 20px;
    }

    .header-text {
      color: $ssb-dark-6;
    }

    &:hover,
    &:focus {
      @include focus-ring;

      .header-text {
        color: $ssb-green-4;
        text-decoration: underline;
        text-underline-position: under;
      }
    }

    &.no-icon {
      text-decoration: underline;
      text-underline-position: under;
    }

    &:disabled {
      .button-grid {
        cursor: not-allowed;

        .header-text {
          color: $ssb-dark-2;
          text-decoration: none;
        }

        .expand-icon * {
          color: $ssb-dark-2;
        }
      }
    }
  }

  .accordion-body {
    @include open-sans;
    color: $ssb-dark-6;
    font-size: 16px;

    &.closed {
      display: none;
      height: 0;
    }
  }

  &.negative {
    .button-header {
      &.no-icon {
        text-decoration: underline;
        text-decoration-color: $ssb-white;
      }

      .button-grid {
        .header-text {
          color: $ssb-white;
        }

        svg {
          color: $ssb-green-2;
        }
      }

      &:disabled {
        .button-grid {
          cursor: not-allowed;

          .header-text {
            color: $ssb-dark-4 !important;
            text-decoration: none;
          }

          svg {
            color: $ssb-dark-4;
          }

          .expand-icon * {
            color: $ssb-dark-4;
          }
        }

        &.no-icon {
          .button-grid .header-text {
            text-decoration: underline;
          }
        }
      }
    }

    .accordion-body {
      color: $ssb-white;
    }
  }
}
