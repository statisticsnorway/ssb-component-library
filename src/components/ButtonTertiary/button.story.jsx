import React from 'react';
import { storiesOf } from '@storybook/react';
import { ChevronDown } from 'react-feather';
import Button from './index';

const negativeStyle = {
	backgroundColor: '#274247',
	padding: '100px',
	display: 'grid',
	gridTemplateColumns: 'repeat(3, auto)',
	gridColumnGap: '20px',
};

storiesOf('Button', module)
	.add('Tertiary', () => (
		<div style={{ display: 'grid', gridTemplateColumns: 'repeat(3, auto)', gridColumnGap: '20px' }}>
			<Button className="mr-4">Tertiary button</Button>
			<Button className="mr-4" icon={<ChevronDown size={18} />}>Tertiary icon</Button>
		</div>
	));
