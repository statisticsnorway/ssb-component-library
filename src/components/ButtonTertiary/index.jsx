import React, { useState } from 'react'
import PropTypes from 'prop-types'

import { ChevronDown, ChevronUp } from 'react-feather'

const ButtonTertiary = ({
  id,
  children,
  className,
  header,
  openByDefault,
  tabIndex,
  accordion,
  disabled,
  negative,
  onClick,
}) => {
  const [isOpen, toggleOpen] = useState(openByDefault)
  const classNames = `ssb-btn-tertiary${negative ? ' negative' : ''}${className ? ` ${className}` : ''}`

  return (
    <div id={id} className={classNames.replace(/\s\s+/g, ' ').trim()}>
      <button
        className={`button-header ${isOpen ? 'open' : 'closed'} ${accordion ? 'icon' : 'no-icon'}`}
        aria-expanded={isOpen ? 'true' : 'false'}
        tabIndex={tabIndex}
        onClick={accordion ? () => toggleOpen(!isOpen) : onClick}
        disabled={disabled}
        id={`accordion-button-${id}`}
        aria-controls={accordion && 'accordion-section'}
        aria-disabled={disabled && 'true'}
      >
        <span className='button-grid'>
          <span className='header-text'>{header}</span>
          {accordion && !isOpen && <ChevronDown className='expand-icon' size={22} />}
          {accordion && isOpen && <ChevronUp className='expand-icon' size={22} />}
        </span>
      </button>
      {accordion && (
        <div
          id='accordion-section'
          className={`accordion-body ${isOpen ? 'open' : 'closed'}`}
          role='region'
          aria-labelledby={`accordion-button-${id}`}
        >
          {children}
        </div>
      )}
    </div>
  )
}

ButtonTertiary.defaultProps = {
  openByDefault: false,
  tabIndex: 0,
  accordion: true,
  disabled: false,
  negative: false,
  onClick: () => {},
}

ButtonTertiary.propTypes = {
  id: PropTypes.string,
  children: PropTypes.node,
  className: PropTypes.string,
  header: PropTypes.string,
  openByDefault: PropTypes.bool,
  tabIndex: PropTypes.number,
  accordion: PropTypes.bool,
  disabled: PropTypes.bool,
  negative: PropTypes.bool,
  onClick: PropTypes.func,
}

export default ButtonTertiary
