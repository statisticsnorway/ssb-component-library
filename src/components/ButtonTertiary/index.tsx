import React, { useState, type ReactNode } from 'react'
import { ChevronDown, ChevronUp } from 'react-feather'

interface ButtonTertiaryProps {
  id?: string
  children?: ReactNode
  className?: string
  header?: string
  openByDefault?: boolean
  tabIndex?: number
  accordion?: boolean
  disabled?: boolean
  negative?: boolean
  onClick?: () => void
}

const ButtonTertiary: React.FC<ButtonTertiaryProps> = ({
  id,
  children,
  className,
  header,
  openByDefault = false,
  tabIndex = 0,
  accordion = true,
  disabled = false,
  negative = false,
  onClick = () => {},
}) => {
  const [isOpen, toggleOpen] = useState(openByDefault)
  const classNames = `ssb-btn-tertiary${negative ? ' negative' : ''}${className ? ` ${className}` : ''}`

  return (
    <div id={id} className={classNames.replace(/\s\s+/g, ' ').trim()}>
      <button
        className={`button-header ${isOpen ? 'open' : 'closed'} ${accordion ? 'icon' : 'no-icon'}`}
        aria-expanded={isOpen ? 'true' : 'false'}
        tabIndex={tabIndex}
        onClick={accordion ? () => toggleOpen(!isOpen) : onClick}
        disabled={disabled}
        id={`accordion-button-${id}`}
        aria-controls={accordion ? 'accordion-section' : undefined}
        aria-disabled={disabled ? 'true' : undefined}
      >
        <span className='button-grid'>
          <span className='header-text'>{header}</span>
          {accordion && !isOpen && <ChevronDown className='expand-icon' size={22} />}
          {accordion && isOpen && <ChevronUp className='expand-icon' size={22} />}
        </span>
      </button>
      {accordion && (
        <div
          id='accordion-section'
          className={`accordion-body ${isOpen ? 'open' : 'closed'}`}
          role='region'
          aria-labelledby={`accordion-button-${id}`}
        >
          {children}
        </div>
      )}
    </div>
  )
}

export default ButtonTertiary
