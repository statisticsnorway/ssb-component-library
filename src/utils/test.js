import fs from 'fs';
import path from 'path';
import { render as rtlRender } from '@testing-library/react';

export * from '@testing-library/react';
export * from '@testing-library/jest-dom';

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8');

export function render(ui, options) {
	const view = rtlRender(ui, options);

	const styleElement = document.createElement('style');
	styleElement.innerHTML = styles;
	document.body.appendChild(styleElement);

	return view;
}
