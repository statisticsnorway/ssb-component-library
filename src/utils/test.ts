import fs from 'fs'
import path from 'path'
import { RenderOptions, RenderResult, render as rtlRender } from '@testing-library/react'
import { ReactNode } from 'react'

export * from '@testing-library/react'

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8')

export function render(ui: ReactNode, options?: RenderOptions): RenderResult {
  const view = rtlRender(ui, options)

  const styleElement = document.createElement('style')
  styleElement.innerHTML = styles
  document.body.appendChild(styleElement)

  return view
}
