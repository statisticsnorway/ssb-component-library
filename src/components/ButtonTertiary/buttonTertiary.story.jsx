import React from 'react'
import ButtonTertiary from './index'
import Card from '../Card'

const testOnClick = () => {
  console.log('Im an onClick function for testing')
}

export default {
  title: 'ButtonTertiary',
}

export const Tertiary = () => (
  <>
    <div
      style={{
        width: '800px',
        height: 'auto',
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'flex-start',
      }}
    >
      <ButtonTertiary className='mr-4' header='Knapp med Icon og Content'>
        <div style={{ display: 'flex', flexDirection: 'inline', gridColumnGap: '11px' }}>
          <Card title='Tittel' href=' '>
            <p>Explain something about something with something clever.</p>
          </Card>
          <Card title='Tittel' href=' '>
            <p>Explain something about something with something clever.</p>
          </Card>
          <Card title='Tittel' href=' '>
            <p>Explain something about something with something clever.</p>
          </Card>
        </div>
      </ButtonTertiary>

      <ButtonTertiary className='mr-4' header='Knapp' accordion={false} onClick={() => testOnClick()} />

      <ButtonTertiary className='mr-4' header='Disabled Knapp' disabled>
        <p>Explain something about something with something clever.</p>
      </ButtonTertiary>
    </div>

    <div
      style={{
        width: '800px',
        height: 'auto',
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'flex-start',
        background: '#224742',
      }}
    >
      <ButtonTertiary className='mr-4' header='Knapp negative' negative accordion={false} onClick={() => testOnClick()}>
        <p>Explain something about something with something clever.</p>
      </ButtonTertiary>

      <ButtonTertiary className='mr-4' header='Knapp negative med Icon' negative>
        <p>Explain something about something with something clever.</p>
      </ButtonTertiary>

      <ButtonTertiary
        className='mr-4'
        header='Knapp negative med Icon'
        accordion={false}
        negative
        disabled
        onClick={() => testOnClick()}
      >
        <p>Explain something about something with something clever.</p>
      </ButtonTertiary>

      <ButtonTertiary className='mr-4' header='Knapp negative med Icon' negative disabled>
        <p>Explain something about something with something clever.</p>
      </ButtonTertiary>
    </div>
  </>
)
