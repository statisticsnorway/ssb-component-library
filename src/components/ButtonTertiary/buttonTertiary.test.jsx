import React from 'react'
import userEvent from '@testing-library/user-event'
import { expect } from '@jest/globals'
import { screen, render } from '../../utils/test'
import ButtonTertiary from './index'

const header = 'Click Me!'

describe('Button Tertiary component', () => {
  test('Matches the snapshot', () => {
    const { asFragment } = render(<ButtonTertiary header={header}>Body</ButtonTertiary>)
    expect(asFragment()).toMatchSnapshot()
  })
  test('Applies negative class correctly', () => {
    const { asFragment } = render(
      <ButtonTertiary header={header} negative={true}>
        Body
      </ButtonTertiary>
    )
    expect(asFragment()).toMatchSnapshot()
  })
  test('Disables button when prop is true', async () => {
    render(
      <ButtonTertiary id={'1'} disabled={true} header={header}>
        Body
      </ButtonTertiary>
    )
    const button = await screen.findByRole('button', { text: header })
    expect(button.disabled).toEqual(true)
  })
  test('Renders correct header text', async () => {
    render(<ButtonTertiary header={header}>Button</ButtonTertiary>)
    const el = await screen.findByText(header)
    expect(el.parentElement).toBeVisible()
  })
  test('Closes after click when starting open', async () => {
    const user = userEvent.setup()
    render(
      <ButtonTertiary id={'1'} openByDefault={true} header={header}>
        Body
      </ButtonTertiary>
    )
    expect(await screen.findByText('Body')).toBeVisible()

    const button = await screen.findByRole('button')
    await user.click(button)

    expect(await screen.findByText('Body')).not.toBeVisible()
  })
  test('Renders toggle icon when accordion prop is true', () => {
    const { asFragment } = render(
      <ButtonTertiary header={header} accordion={true}>
        Body
      </ButtonTertiary>
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
