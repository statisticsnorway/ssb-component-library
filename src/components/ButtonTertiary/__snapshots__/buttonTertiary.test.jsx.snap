// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Tertiary component Applies negative class correctly 1`] = `
<DocumentFragment>
  <div
    class="ssb-btn-tertiary negative"
  >
    <button
      aria-controls="accordion-section"
      aria-disabled="false"
      aria-expanded="false"
      class="button-header closed icon"
      id="accordion-button-undefined"
      tabindex="0"
    >
      <span
        class="button-grid"
      >
        <span
          class="header-text"
        >
          Click Me!
        </span>
        <svg
          class="expand-icon"
          fill="none"
          height="22"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="22"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </span>
    </button>
    <div
      aria-labelledby="accordion-button-undefined"
      class="accordion-body closed"
      id="accordion-section"
      role="region"
    >
      Body
    </div>
  </div>
</DocumentFragment>
`;

exports[`Button Tertiary component Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ssb-btn-tertiary"
  >
    <button
      aria-controls="accordion-section"
      aria-disabled="false"
      aria-expanded="false"
      class="button-header closed icon"
      id="accordion-button-undefined"
      tabindex="0"
    >
      <span
        class="button-grid"
      >
        <span
          class="header-text"
        >
          Click Me!
        </span>
        <svg
          class="expand-icon"
          fill="none"
          height="22"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="22"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </span>
    </button>
    <div
      aria-labelledby="accordion-button-undefined"
      class="accordion-body closed"
      id="accordion-section"
      role="region"
    >
      Body
    </div>
  </div>
</DocumentFragment>
`;

exports[`Button Tertiary component Renders toggle icon when accordion prop is true 1`] = `
<DocumentFragment>
  <div
    class="ssb-btn-tertiary"
  >
    <button
      aria-controls="accordion-section"
      aria-disabled="false"
      aria-expanded="false"
      class="button-header closed icon"
      id="accordion-button-undefined"
      tabindex="0"
    >
      <span
        class="button-grid"
      >
        <span
          class="header-text"
        >
          Click Me!
        </span>
        <svg
          class="expand-icon"
          fill="none"
          height="22"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="22"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </span>
    </button>
    <div
      aria-labelledby="accordion-button-undefined"
      class="accordion-body closed"
      id="accordion-section"
      role="region"
    >
      Body
    </div>
  </div>
</DocumentFragment>
`;
